core/domain/story_services.py

```diff
@@ -289,8 +307,24 @@ def update_story(
         raise ValueError('Expected a commit message but received none.')
 
     old_story = story_fetchers.get_story_by_id(story_id)
-    new_story = apply_change_list(story_id, change_list)
+    new_story, exp_ids_removed_from_story, exp_ids_added_to_story = (
+        apply_change_list(story_id, change_list))
     _save_story(committer_id, new_story, commit_message, change_list)
+
+    exploration_context_models_to_be_deleted = (
+        exp_models.ExplorationContextModel.get_multi(
+            exp_ids_removed_from_story))
+    exploration_context_models_to_be_deleted = [
+        model for model in exploration_context_models_to_be_deleted
+        if model is not None]
+    exp_models.ExplorationContextModel.delete_multi(
+        exploration_context_models_to_be_deleted)
+
+    new_exploration_context_models = [exp_models.ExplorationContextModel(
```

Comment:
Should there be any collision checking done here (i.e. to prevent an exploration getting assigned to multiple stories)?

Reply:
Done